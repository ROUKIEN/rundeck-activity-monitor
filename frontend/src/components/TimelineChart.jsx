import React from 'react'
import { Chart } from 'react-google-charts'
import { parseExecutions } from '../utils'

const Loader = () =>
  <div className="message">
    <div className="message-body">Loading, please wait...</div>
  </div>

const FailedToLoad = () =>
  <div className="message">
    <div className="message-body">Failed to render executions.</div>
  </div>

function TimelineChart({ executions }) {
  const { data } = parseExecutions(executions)

  return (
    <Chart
      chartType="Timeline"
      data={data}
      options={
        {
          timeline: {
            showBarLabels: false
          }
        }
      }
      width="100%"
      height="600px"
      loader={<Loader></Loader>}
      errorElement={<FailedToLoad></FailedToLoad>}
      chartEvents={
        [
          {
            eventName: 'select',
            callback: (e) => {
              const selection = e.chartWrapper.getChart().getSelection()
              const execution = executions[selection[0].row]
              window.open(execution.execution_permalink, '_blank').focus()
            }
          }
        ]
      }
    />
  )
}

export default TimelineChart
