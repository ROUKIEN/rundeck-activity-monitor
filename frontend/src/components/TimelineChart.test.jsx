import React from 'react'
import renderer from 'react-test-renderer'
import { expect, test } from 'vitest'
import TimelineChart from './TimelineChart'

test('TimelineChart renders correctly', () => {
  const executions = [
    {
      id: 142,
      rundeck_instance: 'rundeck_dev',
      rundeck_project: 'my_project',
      rundeck_job: 'my job',
      rundeck_job_id: '339164fc-b1e8-4198-babd-fff0e9055e4c',
      execution_status: 'succeeded',
      execution_id: 1244,
      execution_start: '2022-04-05T09:10:00Z',
      execution_end: '2022-04-05T09:10:01Z',
      execution_permalink: 'https://rundeck.dev/project/my_project/execution/show/1244'
    },
    {
      id: 143,
      rundeck_instance: 'rundeck_dev',
      rundeck_project: 'my_project',
      rundeck_job: 'my job',
      rundeck_job_id: '339164fc-b1e8-4198-babd-fff0e9055e4c',
      execution_status: 'succeeded',
      execution_id: 1245,
      execution_start: '2022-04-05T09:10:00Z',
      execution_end: '2022-04-05T09:10:01Z',
      execution_permalink: 'https://rundeck.dev/project/my_project/execution/show/1245'
    },
  ]
  const tc = renderer.create(
    <TimelineChart executions={executions}></TimelineChart>
  ).toJSON()
  expect(tc).toMatchSnapshot('Rendered list without any facet selected')
})
