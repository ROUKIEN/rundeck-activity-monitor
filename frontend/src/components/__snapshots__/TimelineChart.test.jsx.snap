// Vitest Snapshot v1

exports[`TimelineChart renders correctly > Rendered list without any facet selected 1`] = `
[
  <h3
    className="title is-4"
  >
    3
     executions
  </h3>,
  <div
    className="message"
  >
    <div
      className="message-body"
    >
      Loading, please wait...
    </div>
  </div>,
]
`;
